### PR TITLE
[system] Elaborate tiles in order of hartid

### PR DESCRIPTION
**Pull request by jerryz123**
When the harts are defined out-of-order, the DTS entry for the CLINT will also be misordered. It was difficult to track down where that was occuring. For now we can fix this by defining the tiles by order of hartid.

I believe the new code is also DRYer than the old implementation.

### CODE REVIEW

File: generators/utilities/src/main/scala/Subsystem.scala [hunk 1]
```diff
@@ -34,49 +34,47 @@ trait HasBoomAndRocketTiles extends HasTiles
...
-
-    connectMasterPortsToSBus(rocket, crossing)
-    connectSlavePortsToCBus(rocket, crossing)
-
-    def treeNode: RocketTileLogicalTreeNode = new RocketTileLogicalTreeNode(rocket.rocketLogicalTree.getOMInterruptTargets)
-    LogicalModuleTree.add(logicalTreeNode, rocket.rocketLogicalTree)
-
-    rocket
-  }
+  // This MUST be performed in order of hartid
```
Comment (jwright6323) [inline generators/utilities/src/main/scala/Subsystem.scala]: Probably worth explaining why

File: generators/utilities/src/main/scala/Subsystem.scala [hunk 2]
```diff
@@ -34,49 +34,47 @@ trait HasBoomAndRocketTiles extends HasTiles
...
+  val tiles = allTilesInfo.sortWith(_._1.hartId < _._1.hartId).map {
+    case (param, crossing) => {
+      val (tile, rocketLogicalTree) = param match {
+        case r: RocketTileParams => {
+          val t = LazyModule(new RocketTile(r, crossing, PriorityMuxHartIdFromSeq(rocketTileParams), logicalTreeNode))
+          (t, t.rocketLogicalTree)
+        }
+        case b: BoomTileParams => {
+          val t = LazyModule(new BoomTile(b, crossing, PriorityMuxHartIdFromSeq(boomTileParams), logicalTreeNode))
+          (t, t.rocketLogicalTree) // TODO FIX rocketLogicalTree is not a member of the superclass, both child classes define it separately
```
Comment (abejgonzalez) [inline generators/utilities/src/main/scala/Subsystem.scala]: Is there any way to reasonably fix this without a change to RC?

**Comment by abejgonzalez** [comment]
LGTM. I still want to test this with a Linux boot. Additionally, we should wait until CI passes on it.

File: generators/utilities/src/main/scala/Subsystem.scala (hunk 2 above)
Comment (jerryz123) [reply generators/utilities/src/main/scala/Subsystem.scala]: Need to fix the LogicalTree representation for BOOM to not just use the Rocket one.

**Comment by abejgonzalez** [comment]
I am fixing the error in firesim right now. I will push a fix once Linux works.

**Comment by abejgonzalez** [comment]
Testing this as we speak on FireSim with a SmallBoomConfig and SmallBoom+RocketConfig

File: generators/utilities/src/main/scala/Subsystem.scala (hunk 1 above)
Comment (jerryz123) [reply generators/utilities/src/main/scala/Subsystem.scala]:
```suggestion
  // This MUST be performed in order of hartid
  // There is something weird with registering tile-local interrupt controllers to the CLINT.
  // TODO: investigate why
```

File: generators/firechip/src/main/scala/TargetMixins.scala [hunk 3]
```diff
@@ -9,12 +9,15 @@ import freechips.rocketchip.tilelink._
 import freechips.rocketchip.subsystem._
 import freechips.rocketchip.amba.axi4._
 import freechips.rocketchip.util._
+import freechips.rocketchip.tile._
```
Comment (jerryz123) [inline generators/firechip/src/main/scala/TargetMixins.scala]:
```suggestion
import freechips.rocketchip.tile.RocketTile
```

**Comment by jerryz123** [comment]
Superseded by #277

**Pull request closed without merging**